File: apis/backend-for-frontend/src/channels/rest/express/controllers/translators.ts
import { Genre } from '@constants/index';
import Album from '@entities/Album';
import Artist from '@entities/Artist';
import Music from '@entities/Music';
import Playlist, { Track } from '@entities/Playlist';
import { getAlbumDuration } from '@utils/index';

export function translateArtist(artist: Artist): any {
  return {
    id: artist.id,
    name: artist.name,
    country: artist.country,
    foundation_date: artist.foundationDate,
    members: artist.members,
    description: artist.description,
    genre: Genre[artist.genre],
    photos: artist.photos,
    profile_img: artist.photos[0],
    full_img: artist.photos.length > 1 ? artist.photos[1] : '',
    vertical_img: artist.photos.length > 2 ? artist.photos[2] : '',
    facebook_url: artist.facebookUrl,
    twitter_url: artist.twitterUrl,
    instagram_url: artist.instagramUrl,
    wikipedia_url: artist.wikipediaUrl,
    font: artist.font,
    favorites: artist.favorites,
    followers: artist.followers,
    albums: artist.albums.map(album => translateAlbum(album)),
    popular_tracks: artist.popularTracks.map(track => translateMusic(track)),
  };
}

export function translateArtists(artists: Artist[]): any {
  return artists.map(artist => translateArtist(artist));
}

export function translateAlbum(album: Album): any {
  return {
    id: album.id,
    name: album.name,
    release_date: album.releaseDate,
    cover: album.cover,
    studio: album.studio,
    producers: album.producers,
    artist_id: album.artistId,
    tracks: album.tracks.map(track => translateMusic(track)),
    number_of_tracks: album.tracks.length,
    full_duration: getAlbumDuration(album.tracks),
  };
}

export function translateAlbums(albums: Album[]): any {
  return albums.map(album => translateAlbum(album));
}

export function translateMusic(music: Music): any {
  return {
    id: music.id,
    title: music.title,
    duration: music.durationInSeconds,
    file: music.file,
    composers: music.composers,
    lyrics: music.lyrics,
    album_id: music.albumId,
    artist_id: music.artistId,
    views: music.views,
  };
}

export function translatePlaylist(playlist: Playlist): any {
  return {
    id: playlist.id,
    name: playlist.name,
    user_id: playlist.userId,
    tracks: playlist.tracks.map(track => translatePlaylistTrack(track)),
  };
}

export function translatePlaylists(playlists: Playlist[]): any {
  return playlists.map(playlist => translatePlaylist(playlist));
}

export function translatePlaylistTrack(track: Track): any {
  return {
    id: track.id,
    index: track.index,
    music: track.music ? translateMusic(track.music) : null,
  };
}

File: apis/users-musics-service/internal/integrations/musicsIntegration/translators.go
package musicsIntegration

import (
	e "users-musics-service/internal/entities"
	"users-musics-service/internal/integrations/musicsIntegration/proto"
)

func translateArtistEntity(artist *proto.Artist) *e.Artist {
	return &e.Artist{
		Id:           artist.Id,
		Name:         artist.Name,
		Country:      artist.Country,
		Members:      artist.Members,
		Description:  artist.Description,
		Photos:       artist.Photos,
		FacebookUrl:  artist.FacebookUrl,
		TwitterUrl:   artist.TwitterUrl,
		InstagramUrl: artist.InstagramUrl,
		WikipediaUrl: artist.WikipediaUrl,
		Font:         artist.Font,
		Favorites:    artist.Favorites,
		Followers:    artist.Followers,
	}
}

func translateMusicEntity(music *proto.Music) *e.Music {
	return &e.Music{
		Id:        music.Id,
		Title:     music.Title,
		Duration:  music.Duration,
		File:      music.File,
		Composers: music.Composers,
		Lyrics:    music.Lyrics,
		Views:     music.Views,
		AlbumId:   music.AlbumId,
		ArtistId:  music.ArtistId,
	}
}
